### ISSUE TITLE
scipy.spatial.SphericalVoronoi gives incorrect results

### ISSUE THREAD

**Issue by rtshort** [scipy/scipy#9386]
When all of the points are in one hemisphere some of the Voronoi points are incorrect.

To be more precise, certain of the points are on the wrong side of the sphere.  An example is to create a tetrahedron with the following generator points:

generators (polar)
 [[-25.56198973  10.81015471]
 [-54.40661695  26.73672285]
 [-55.29180524  -6.7928405 ]
 [-45.12839961   9.12243389]]
generators (rectangular)
 [[ 0.88610999 -0.42383021  0.18755541]
 [ 0.51980039 -0.72622668  0.4498915 ]
 [ 0.56540011 -0.81629197 -0.11827989]
 [ 0.69659682 -0.69972598  0.15854467]]

![p](https://user-images.githubusercontent.com/2798381/47086102-7b2f5300-d1cd-11e8-9394-405ca519bca6.png)

The first three Voronoi points and associated Delaunay triangles come out fine:
Voronoi vertex number 1 , coordinates: [ 0.73647759 -0.54764287  0.39709955]
Delaunay vertices
 [[ 0.69659682 -0.69972598  0.15854467]
 [ 0.88610999 -0.42383021  0.18755541]
 [ 0.51980039 -0.72622668  0.4498915 ]]
Voronoi vertex number 2 , coordinates: [ 0.42155897 -0.88977832  0.17487877]
Delaunay vertices
 [[ 0.69659682 -0.69972598  0.15854467]
 [ 0.56540011 -0.81629197 -0.11827989]
 [ 0.51980039 -0.72622668  0.4498915 ]]
Voronoi vertex number 3 , coordinates: [ 0.82153008 -0.54761634 -0.15875979]
Delaunay vertices
 [[ 0.69659682 -0.69972598  0.15854467]
 [ 0.56540011 -0.81629197 -0.11827989]
 [ 0.88610999 -0.42383021  0.18755541]]

![v1](https://user-images.githubusercontent.com/2798381/47086240-ce090a80-d1cd-11e8-83be-95237aff6c81.png)
![v2](https://user-images.githubusercontent.com/2798381/47086249-d2352800-d1cd-11e8-8215-32c9c047f499.png)
![v3](https://user-images.githubusercontent.com/2798381/47086267-de20ea00-d1cd-11e8-945c-1280f73d67a4.png)

The fourth point is not correct - it should be on the opposite side of the sphere:
Voronoi vertex number 0 , coordinates: [ 0.69593711 -0.69851441  0.16658079]
Delaunay vertices
 [[ 0.56540011 -0.81629197 -0.11827989]
 [ 0.88610999 -0.42383021  0.18755541]
 [ 0.51980039 -0.72622668  0.4498915 ]]

![v0](https://user-images.githubusercontent.com/2798381/47086358-17f1f080-d1ce-11e8-862c-1902bb002aca.png)

I have a good understanding of the algorithm and what the problem is but I don't know how to fix it using qhull.

**Comment by tylerjereddy** [comment]
Related / duplicate of #8859 & also mentioned in #5232.

I've read a few of the algorithm papers in the (often recent) literature, and haven't seen one mention a proof for a working algorithm with single hemisphere input.

> I have a good understanding of the algorithm and what the problem is but I don't know how to fix it using qhull.

`SphericalVoronoi` is not implemented using `qhull` under the hood, unlike many other components of the computational geometry infrastructure in SciPy. It is mostly Python with a bit of Cython mixed in.

I suspect a robust PR or suggestion of a citation we could follow to solve this issue would be most welcome. cc @WWmore, @fietew, and @niknow.

We may have to be a little cautious if the algorithm isn't published in some form though--I think SciPy generally aims for robust / established algorithms (indeed, most of the Spherical Voronoi papers probably just barely pass that threshold).

**Comment by tylerjereddy** [comment]
That said, we probably do call 1 or 2 qhull functions for things like convex hull or similar in the current algorithm

**Comment by rtshort** [comment]
Sorry about the duplicate issue.  I did search but maybe my github skills are a little weak.

I thought the spherical Voronoi algorithm uses scipy.spatial.ConvexHull to compute the convex hull, and the documentation for that algorithm clearly says it uses QHull to compute the convex hull.  BTW, I am not a geometer.  My math and computer skills are pretty good and I have done a lot of reading on this topic but I am certainly not a professional geometer.

The problem is that ConvexHull doesn't produce the right convex hull.  It computes the hull as a set of points in 3-space, but the convex hull as discussed in Caroli and the other references is a set of regions on the sphere.  The 3d hull can give you the spherical hull but you need to partition the 3d hull into the outer hull and the inner hull (this is the part I don't know how to do).  Projecting the circumcenters of the outer hull onto the sphere as it is currently done gives the "correct" part of the tesselation.  Projecting the circumcenters of the inner hull to the OPPOSITE side of the sphere gives the correct tesselation.

The Ph.D. thesis by Kevin Q. Brown, "Geometric Transforms for Fast Geometric Algorithms", 1979 and the masters thesis "Construction of Delaunay Triangulations on the Sphere: A Parallel Approach" by Veronica Larrea (especially the latter) give algorithms, although proofs are lacking in both cases.

**Comment by rtshort** [comment]
Just posted as you were posting.  Unless the code has changed since I downloaded it, it calls ConvexHull to do the hull.

**Comment by tylerjereddy** [comment]
@rtshort Yeah, you're right about that -- we didn't re-write the convex hull algorithm from scratch :)

Those are some interesting insights! The Brown thesis is indeed pretty famous, but your suggestions about splitting the 3D hull into an outer and inner hull are intriguing!

**Comment by rtshort** [comment]
The Larrea thesis was more insightful for me.  That was where I got the aha! about the spherical hull vs. the 3d hull and the notion of inner and outer hull.  The notion of using the inner and outer 3d hulls is mine (although I doubt it is original).  It is a lot easier to visualize in 2d.  Looking at the following picture I have three points on the circle (2-sphere).  The figure to the right is the 2d convex hull, i.e. the hull computed by taking the points in 2-space.  The figure on the bottom shows the convex hull on the circle, in which each set of points defines an arc of the hull (colored).  The center of each arc are the Voronoi points and can be found by projecting from the 2d hull segments to the circle, but the key is that the inner hull (always a single segment in a 2-sphere) projects to the opposite side.   I sketched out a proof for this, but have not tried to fill it in.

All I wanted was to find a point in a geographic region that was furthest from the boundaries.  Just for my own edification.  Instead I spend many weeks learning computational geometry!  Actually pretty interesting and fun in a geeky sort of way.

![bug](https://user-images.githubusercontent.com/2798381/47168274-531b1f00-d2b5-11e8-8d87-fbf3ce708262.png)
